src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfoCacheService.cs

```diff
@@ -59,138 +91,124 @@ public SymbolTreeInfoCacheService()
-                _projectIdToInfo[document.Project.Id] = newInfo;
-                return;
+                // NOTE: currently we process projects serially.  This is because the same metadata reference might be
+                // found in multiple projects and we can't currently process that in parallel.
+                await AnalyzeProjectAsync(project, cancellationToken).ConfigureAwait(false);
             }
 
-            await AnalyzeProjectAsync(document.Project, cancellationToken).ConfigureAwait(false);
+            // Now that we've produced all the indices for the projects asked for, also remove any indices for projects
+            // no longer in the solution.
+            var removedProjectIds = _projectIdToInfo.Keys.Except(solution.ProjectIds).ToArray();
+            foreach (var projectId in removedProjectIds)
+                this.RemoveProject(projectId);
         }
 
-        public async Task AnalyzeProjectAsync(Project project, CancellationToken cancellationToken)
+        private async Task AnalyzeProjectAsync(Project project, CancellationToken cancellationToken)
         {
-            if (!project.SupportsCompilation)
-                return;
-
-            // Produce the indices for the source and metadata symbols in parallel.
             using var _ = ArrayBuilder<Task>.GetInstance(out var tasks);
 
-            tasks.Add(Task.Run(() => this.UpdateSourceSymbolTreeInfoAsync(project, cancellationToken), cancellationToken));
-            tasks.Add(Task.Run(() => this.UpdateReferencesAsync(project, cancellationToken), cancellationToken));
+            // We can compute source-symbols in parallel with the metadata symbols.
+            tasks.Add(CreateWorkAsync(() => this.UpdateSourceSymbolTreeInfoAsync(project, cancellationToken), cancellationToken));
+
+            // Add tasks to update the symboltree for all metadata references.  As these are all distinct, they can run
+            // in parallel as we won't be trying to update the associated data for the same reference at the same time.
+            foreach (var reference in project.MetadataReferences.OfType<PortableExecutableReference>().Distinct())
+                tasks.Add(CreateWorkAsync(() => UpdateReferenceAsync(project, reference, cancellationToken), cancellationToken));
 
             await Task.WhenAll(tasks).ConfigureAwait(false);
         }
 
         private async Task UpdateSourceSymbolTreeInfoAsync(Project project, CancellationToken cancellationToken)
         {
-            var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(project, cancellationToken).ConfigureAwait(false);
+            // Find the top-level-semantic-version of this project.  We only want to recompute if it has changed. This is
+            // because the symboltree contains the names of the types/namespaces in the project and would not change
+            // if the semantic-version of the project hasn't changed.  We also do not need to check the 'dependent
```

Comment:
remind me - does the semantic version change when private members change?  iirc only 'visible' changes would be included but I might not be remembering right.

Reply:
Yes.  this calls `newTree.IsEquivalentTo(oldTree, topLevel: true)` which will consider privates when checking things.